File: Usage.cs
using System.Globalization;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text.Json;

namespace ClaudeUsageBar;

/// <summary>One usage limit. Length is the window's duration (5 hours, 7 days), used for labels and the pace marker.</summary>
public sealed record UsageWindow(double Percent, DateTimeOffset? ResetsAt, TimeSpan? Length = null);
public sealed record ExtraUsage(bool Enabled, double? LimitUsd, double? UsedUsd, double? Percent);

public sealed class UsageSnapshot
{
    public UsageWindow? FiveHour { get; init; }
    public UsageWindow? SevenDay { get; init; }
    public UsageWindow? SevenDayOpus { get; init; }
    public UsageWindow? SevenDaySonnet { get; init; }
    public ExtraUsage? Extra { get; init; }
    public DateTimeOffset FetchedAt { get; init; }

    public static UsageSnapshot Parse(JsonElement root, DateTimeOffset fetchedAt) => new()
    {
        FiveHour = Window(root, "five_hour", Fmt.FiveHours),
        SevenDay = Window(root, "seven_day", Fmt.SevenDays),
        SevenDayOpus = Window(root, "seven_day_opus", Fmt.SevenDays),
        SevenDaySonnet = Window(root, "seven_day_sonnet", Fmt.SevenDays),
        Extra = ParseExtra(root),
        FetchedAt = fetchedAt,
    };

    /// <summary>
    /// Codex's usage response: rate_limit.primary_window (the short one, 5 hours) goes in the session slot and
    /// secondary_window (weekly) in the weekly slot. Percent is used_percent; reset_at is Unix seconds.
    /// </summary>
    public static UsageSnapshot ParseChatGpt(JsonElement root, DateTimeOffset fetchedAt)
    {
        JsonElement rl = default;
        bool has = root.ValueKind == JsonValueKind.Object && root.TryGetProperty("rate_limit", out rl) && rl.ValueKind == JsonValueKind.Object;
        return new()
        {
            FiveHour = has ? CodexWindow(rl, "primary_window", fetchedAt) : null,
            SevenDay = has ? CodexWindow(rl, "secondary_window", fetchedAt) : null,
            FetchedAt = fetchedAt,
        };
    }

    static UsageWindow? CodexWindow(JsonElement rl, string name, DateTimeOffset fetchedAt)
    {
        if (!rl.TryGetProperty(name, out var w) || w.ValueKind != JsonValueKind.Object) return null;
        var pct = Num(w, "used_percent");
        if (pct is null) return null;
        DateTimeOffset? reset = Num(w, "reset_at") is { } at && at > 0 ? DateTimeOffset.FromUnixTimeSeconds((long)at)
            : Num(w, "reset_after_seconds") is { } after ? fetchedAt.AddSeconds(after) : null;
        TimeSpan? length = Num(w, "limit_window_seconds") is { } secs && secs > 0 ? TimeSpan.FromSeconds(secs) : null;
        return new UsageWindow(Math.Clamp(pct.Value, 0, 100), reset, length);
    }

    static UsageWindow? Window(JsonElement root, string name, TimeSpan length)
    {
        if (root.ValueKind != JsonValueKind.Object || !root.TryGetProperty(name, out var w) || w.ValueKind != JsonValueKind.Object)
            return null;
        var pct = Num(w, "utilization");
        if (pct is null) return null;
        DateTimeOffset? reset = null;
        if (w.TryGetProperty("resets_at", out var r) && r.ValueKind == JsonValueKind.String &&
            DateTimeOffset.TryParse(r.GetString(), CultureInfo.InvariantCulture, DateTimeStyles.AssumeUniversal, out var parsed))
            reset = parsed;
        return new UsageWindow(Math.Clamp(pct.Value, 0, 100), reset, length);
    }

    static ExtraUsage? ParseExtra(JsonElement root)
    {
        if (!root.TryGetProperty("extra_usage", out var e) || e.ValueKind != JsonValueKind.Object) return null;
        bool enabled = e.TryGetProperty("is_enabled", out var en) && en.ValueKind == JsonValueKind.True;
        // Credits are reported in cents.
        var limit = Num(e, "monthly_limit") / 100.0;
        var used = Num(e, "used_credits") / 100.0;
        var pct = Num(e, "utilization") ?? (limit > 0 && used is not null ? used / limit * 100 : null);
        return new ExtraUsage(enabled, limit, used, pct is null ? null : Math.Clamp(pct.Value, 0, 100));
    }

    static double? Num(JsonElement o, string name) =>
        o.TryGetProperty(name, out var v) && v.ValueKind == JsonValueKind.Number ? v.GetDouble() : null;
}

public enum FetchStatus { None, Ok, NoCredentials, Unauthorized, RateLimited, Error }

public sealed record FetchResult(FetchStatus Status, UsageSnapshot? Data, string? Plan, string? Message = null, TimeSpan? RetryAfter = null);

/// <summary>Whose plan usage the app shows. Picked from the right-click menu.</summary>
public enum Provider { Claude, ChatGpt }

/// <summary>
/// Fetches plan usage with a CLI's own login token (Claude Code or Codex). Subclasses say where the token lives,
/// how to ask for usage and how to read the answer; polling, errors and caching are shared.
/// </summary>
public abstract class UsageClient
{
    // Redirects are refused so the bearer token can only ever go to the provider's host, and responses are size-capped.
    static readonly HttpClient Http = new(new SocketsHttpHandler { AllowAutoRedirect = false, UseCookies = false })
    {
        Timeout = TimeSpan.FromSeconds(20),
        MaxResponseContentBufferSize = 256 * 1024,
    };
    string? _rejectedToken;

    public static UsageClient For(Provider p) => p == Provider.ChatGpt ? new ChatGptClient() : new ClaudeClient();

    public abstract Provider Provider { get; }
    /// <summary>Host the token is sent to, for error messages.</summary>
    protected abstract string Host { get; }
    protected abstract string Company { get; }
    protected abstract string CacheFile { get; }
    protected abstract string[] CliNames { get; }
    /// <summary>Install folders to check besides PATH.</summary>
    protected abstract string[] CliDirs { get; }
    protected abstract (string? token, string? plan, string? account) ReadCredentials();
    protected abstract HttpRequestMessage Request(string token, string? account);
    protected abstract UsageSnapshot Parse(JsonElement root, DateTimeOffset fetchedAt);
    /// <summary>Plan named in the usage response, when the credentials don't say.</summary>
    protected virtual string? PlanFrom(JsonElement root) => null;

    protected static string Home => Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);

    /// <summary>Path to the provider's CLI, or null if it isn't installed.</summary>
    public string? FindCli()
    {
        var dirs = (Environment.GetEnvironmentVariable("PATH") ?? "").Split(';', StringSplitOptions.RemoveEmptyEntries).Concat(CliDirs);
        foreach (var raw in dirs)
        {
            var dir = raw.Trim().Trim('"');
            // Relative PATH entries (e.g. ".") would resolve against our working directory; never launch from those.
            if (!Path.IsPathFullyQualified(dir)) continue;
            foreach (var name in CliNames)
            {
                try { var p = Path.Combine(dir, name); if (File.Exists(p)) return p; } catch { }
            }
        }
        return null;
    }

    string CachePath => Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "ClaudeUsageBar", CacheFile);

    public async Task<FetchResult> FetchAsync()
    {
        var (token, plan, account) = ReadCredentials();
        if (string.IsNullOrEmpty(token))
            return new(FetchStatus.NoCredentials, null, plan, "Not signed in");
        // Don't keep hitting the API with a token it already rejected; wait until the CLI writes a new one.
        if (token == _rejectedToken)
            return new(FetchStatus.Unauthorized, null, plan, "Login expired");

        try
        {
            using var req = Request(token, account);
            req.Headers.UserAgent.ParseAdd("ClaudeUsageBar/1.0");
            using var res = await Http.SendAsync(req);
            var body = await res.Content.ReadAsStringAsync();

            if (res.StatusCode == HttpStatusCode.TooManyRequests)
            {
                TimeSpan? wait = res.Headers.RetryAfter?.Delta
                    ?? (res.Headers.RetryAfter?.Date is { } d ? d - DateTimeOffset.UtcNow : null);
                return new(FetchStatus.RateLimited, null, plan, $"Rate limited by {Company}", wait);
            }
            if (res.StatusCode is HttpStatusCode.Unauthorized or HttpStatusCode.Forbidden)
            {
                _rejectedToken = token;
                return new(FetchStatus.Unauthorized, null, plan, "Login expired");
            }
            if (!res.IsSuccessStatusCode)
                return new(FetchStatus.Error, null, plan, $"Server error ({(int)res.StatusCode})");

            var now = DateTimeOffset.Now;
            using var doc = JsonDocument.Parse(body);
            plan = PlanFrom(doc.RootElement) ?? plan;
            var snap = Parse(doc.RootElement, now);
            SaveCache(body, plan, now);
            return new(FetchStatus.Ok, snap, plan);
        }
        catch (TaskCanceledException) { return new(FetchStatus.Error, null, plan, "Request timed out"); }
        catch (HttpRequestException) { return new(FetchStatus.Error, null, plan, $"Can't reach {Host}"); }
        catch (JsonException) { return new(FetchStatus.Error, null, plan, "Unexpected response"); }
        catch (Exception) { return new(FetchStatus.Error, null, plan, "Couldn't update"); }
    }

    void SaveCache(string rawJson, string? plan, DateTimeOffset at)
    {
        try
        {
            Directory.CreateDirectory(Path.GetDirectoryName(CachePath)!);
            File.WriteAllText(CachePath,
                $"{{\"fetchedAt\":{JsonSerializer.Serialize(at)},\"plan\":{JsonSerializer.Serialize(plan)},\"data\":{rawJson}}}");
        }
        catch { }
    }

    public (UsageSnapshot? data, string? plan) LoadCache()
    {
        try
        {
            using var doc = JsonDocument.Parse(File.ReadAllText(CachePath));
            var root = doc.RootElement;
            var at = root.GetProperty("fetchedAt").GetDateTimeOffset();
            string? plan = root.TryGetProperty("plan", out var p) && p.ValueKind == JsonValueKind.String ? p.GetString() : null;
            return (Parse(root.GetProperty("data"), at), plan);
        }
        catch { return (null, ReadCredentials().plan); }
    }
}

/// <summary>Talks to the same endpoint Claude Code's /usage uses, with Claude Code's own OAuth token.</summary>
public sealed class ClaudeClient : UsageClient
{
    const string Endpoint = "https://api.anthropic.com/api/oauth/usage";

    public override Provider Provider => Provider.Claude;
    protected override string Host => "api.anthropic.com";
    protected override string Company => "Anthropic";
    protected override string CacheFile => "last.json";
    protected override string[] CliNames => new[] { "claude.exe", "claude.cmd" };
    protected override string[] CliDirs => new[] { Path.Combine(Home, ".local", "bin") };

    static string ConfigDir => Environment.GetEnvironmentVariable("CLAUDE_CONFIG_DIR") is { Length: > 0 } d
        ? d : Path.Combine(Home, ".claude");
    static string CredentialsPath => Path.Combine(ConfigDir, ".credentials.json");

    protected override (string? token, string? plan, string? account) ReadCredentials()
    {
        try
        {
            using var doc = JsonDocument.Parse(File.ReadAllText(CredentialsPath));
            if (!doc.RootElement.TryGetProperty("claudeAiOauth", out var o)) return (null, null, null);
            string? token = o.TryGetProperty("accessToken", out var t) ? t.GetString() : null;
            string? plan = o.TryGetProperty("subscriptionType", out var p) && p.ValueKind == JsonValueKind.String ? p.GetString() : null;
            return (token, plan, null);
        }
        catch { return (null, null, null); }
    }

    protected override HttpRequestMessage Request(string token, string? account)
    {
        var req = new HttpRequestMessage(HttpMethod.Get, Endpoint);
        req.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
        req.Headers.Add("anthropic-beta", "oauth-2025-04-20");
        return req;
    }

    protected override UsageSnapshot Parse(JsonElement root, DateTimeOffset fetchedAt) => UsageSnapshot.Parse(root, fetchedAt);
}

/// <summary>
/// Talks to the endpoint Codex's /status uses, with the ChatGPT login Codex stores in ~/.codex/auth.json.
/// These are the Codex limits of the ChatGPT plan: a short (5-hour) and a weekly window.
/// </summary>
public sealed class ChatGptClient : UsageClient
{
    const string Endpoint = "https://chatgpt.com/backend-api/wham/usage";

    public override Provider Provider => Provider.ChatGpt;
    protected override string Host => "chatgpt.com";
    protected override string Company => "OpenAI";
    protected override string CacheFile => "last-chatgpt.json";
    protected override string[] CliNames => new[] { "codex.exe", "codex.cmd" };
    protected override string[] CliDirs => new[]
    {
        Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "npm"),
        Path.Combine(Home, ".local", "bin"),
    };

    static string CodexHome => Environment.GetEnvironmentVariable("CODEX_HOME") is { Length: > 0 } d
        ? d : Path.Combine(Home, ".codex");
    static string AuthPath => Path.Combine(CodexHome, "auth.json");

    protected override (string? token, string? plan, string? account) ReadCredentials()
    {
        try
        {
            using var doc = JsonDocument.Parse(File.ReadAllText(AuthPath));
            if (!doc.RootElement.TryGetProperty("tokens", out var t) || t.ValueKind != JsonValueKind.Object) return (null, null, null);
            string? token = t.TryGetProperty("access_token", out var a) && a.ValueKind == JsonValueKind.String ? a.GetString() : null;
            string? account = t.TryGetProperty("account_id", out var id) && id.ValueKind == JsonValueKind.String ? id.GetString() : null;
            return (token, null, account);
        }
        catch { return (null, null, null); }
    }

    protected override HttpRequestMessage Request(string token, string? account)
    {
        var req = new HttpRequestMessage(HttpMethod.Get, Endpoint);
        req.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
        if (!string.IsNullOrEmpty(account)) req.Headers.TryAddWithoutValidation("ChatGPT-Account-Id", account);
        return req;
    }

    protected override string? PlanFrom(JsonElement root) =>
        root.ValueKind == JsonValueKind.Object && root.TryGetProperty("plan_type", out var p) && p.ValueKind == JsonValueKind.String ? p.GetString() : null;

    protected override UsageSnapshot Parse(JsonElement root, DateTimeOffset fetchedAt) => UsageSnapshot.ParseChatGpt(root, fetchedAt);
}

/// <summary>Everything the UI needs to draw itself.</summary>
public sealed class ViewState
{
    public Provider Provider;
    public UsageSnapshot? Data;
    public string? Plan;
    public FetchStatus Status;
    public string? Message;
    public DateTimeOffset? NextAttempt;
    public bool Fetching;
    public bool CliInstalled = true;
    public TimeSpan PollInterval = TimeSpan.FromMinutes(5);
}

static class Fmt
{
    public static readonly TimeSpan FiveHours = TimeSpan.FromHours(5), SevenDays = TimeSpan.FromDays(7);

    /// <summary>Compact countdown for the taskbar: 42m, 3h 5m, 4d 3h.</summary>
    public static string Short(TimeSpan t)
    {
        if (t <= TimeSpan.Zero) return "now";
        if (t.TotalHours < 1) return $"{Math.Max(1, (int)Math.Ceiling(t.TotalMinutes))}m";
        if (t.TotalDays < 1) return t.Minutes == 0 ? $"{(int)t.TotalHours}h" : $"{(int)t.TotalHours}h {t.Minutes}m";
        return t.Hours == 0 ? $"{(int)t.TotalDays}d" : $"{(int)t.TotalDays}d {t.Hours}h";
    }

    public static string Reset(DateTimeOffset? reset, DateTimeOffset now)
    {
        if (reset is null) return "";
        var left = reset.Value - now;
        if (left <= TimeSpan.Zero) return "Resetting…";
        if (left.TotalHours < 24) return "Resets in " + (left.TotalHours < 1 ? $"{Math.Max(1, (int)Math.Ceiling(left.TotalMinutes))}m" : $"{(int)left.TotalHours}h {left.Minutes:00}m");
        var local = reset.Value.ToLocalTime();
        return "Resets " + local.ToString("ddd", CultureInfo.CurrentCulture) + " " + local.ToString("t", CultureInfo.CurrentCulture);
    }

    public static string Ago(TimeSpan t)
    {
        if (t.TotalSeconds < 45) return "just now";
        if (t.TotalMinutes < 60) return $"{Math.Max(1, (int)Math.Round(t.TotalMinutes))} min ago";
        if (t.TotalHours < 24) return $"{(int)t.TotalHours} h ago";
        return $"{(int)t.TotalDays} d ago";
    }

    public static string In(TimeSpan t) =>
        t.TotalSeconds < 60 ? $"{Math.Max(1, (int)Math.Ceiling(t.TotalSeconds))}s" : $"{(int)Math.Ceiling(t.TotalMinutes)} min";

    public static string Money(double v) => v >= 1000 ? v.ToString("$#,0", CultureInfo.InvariantCulture) : v.ToString("$0.00", CultureInfo.InvariantCulture);

    public static string Plan(string? plan) => plan?.ToLowerInvariant() switch
    {
        null or "" => "",
        "pro" => "Pro",
        "plus" => "Plus",
        "max" => "Max",
        "team" => "Team",
        "enterprise" => "Enterprise",
        var p => CultureInfo.InvariantCulture.TextInfo.ToTitleCase(p),
    };

    /// <summary>Short strip label for a window: "5h", "7d". Falls back when the length is unknown.</summary>
    public static string Label(UsageWindow? w, string fallback) => w?.Length switch
    {
        { TotalDays: >= 1 } l => $"{Math.Round(l.TotalDays):0}d",
        { TotalHours: >= 1 } l => $"{Math.Round(l.TotalHours):0}h",
        _ => fallback,
    };

    /// <summary>"5-hour rolling window", "Weekly window": describes a window by its length.</summary>
    public static string Describe(TimeSpan? length, string fallback) => length switch
    {
        { TotalDays: > 6.5 and < 7.5 } => "Weekly window",
        { TotalDays: >= 1 } l => $"{Math.Round(l.TotalDays):0}-day window",
        { TotalHours: >= 1 } l => $"{Math.Round(l.TotalHours):0}-hour rolling window",
        _ => fallback,
    };

    /// <summary>How far through the window we are (0..1), for the pace marker.</summary>
    public static float? Elapsed(DateTimeOffset? reset, TimeSpan length, DateTimeOffset now)
    {
        if (reset is null) return null;
        var left = reset.Value - now;
        return (float)Math.Clamp(1 - left.TotalSeconds / length.TotalSeconds, 0, 1);
    }
}

File: macos/Sources/ClaudeUsageBar/Usage.swift
import Foundation
import Security

// MARK: - Model

/// Whose plan usage the app shows. Picked from the right-click menu.
enum Provider: String, CaseIterable {
    case claude, chatgpt

    var name: String { self == .chatgpt ? "ChatGPT" : "Claude" }

    static var current: Provider {
        get { UserDefaults.standard.string(forKey: "provider").flatMap(Provider.init(rawValue:)) ?? .claude }
        set { UserDefaults.standard.set(newValue.rawValue, forKey: "provider") }
    }
}

struct UsageWindow: Equatable {
    let percent: Double
    let resetsAt: Date?
    /// The window's duration (5 hours, 7 days), used for labels and the pace marker.
    var length: TimeInterval? = nil
}

struct ExtraUsage: Equatable {
    let enabled: Bool
    let limitUSD: Double?
    let usedUSD: Double?
    let percent: Double?
}

struct UsageSnapshot: Equatable {
    var fiveHour: UsageWindow?
    var sevenDay: UsageWindow?
    var sevenDayOpus: UsageWindow?
    var sevenDaySonnet: UsageWindow?
    var extra: ExtraUsage?
    var fetchedAt: Date

    static func parse(_ root: [String: Any], fetchedAt: Date) -> UsageSnapshot {
        UsageSnapshot(
            fiveHour: window(root["five_hour"], length: Fmt.fiveHours),
            sevenDay: window(root["seven_day"], length: Fmt.sevenDays),
            sevenDayOpus: window(root["seven_day_opus"], length: Fmt.sevenDays),
            sevenDaySonnet: window(root["seven_day_sonnet"], length: Fmt.sevenDays),
            extra: parseExtra(root["extra_usage"]),
            fetchedAt: fetchedAt
        )
    }

    /// Codex's usage response: rate_limit.primary_window (the short one, 5 hours) goes in the session slot and
    /// secondary_window (weekly) in the weekly slot. Percent is used_percent; reset_at is Unix seconds.
    static func parseChatGPT(_ root: [String: Any], fetchedAt: Date) -> UsageSnapshot {
        let rl = root["rate_limit"] as? [String: Any]
        return UsageSnapshot(
            fiveHour: codexWindow(rl?["primary_window"], fetchedAt: fetchedAt),
            sevenDay: codexWindow(rl?["secondary_window"], fetchedAt: fetchedAt),
            fetchedAt: fetchedAt
        )
    }

    static func parse(_ root: [String: Any], provider: Provider, fetchedAt: Date) -> UsageSnapshot {
        provider == .chatgpt ? parseChatGPT(root, fetchedAt: fetchedAt) : parse(root, fetchedAt: fetchedAt)
    }

    private static func codexWindow(_ value: Any?, fetchedAt: Date) -> UsageWindow? {
        guard let o = value as? [String: Any], let pct = number(o["used_percent"]) else { return nil }
        var reset: Date?
        if let at = number(o["reset_at"]), at > 0 { reset = Date(timeIntervalSince1970: at) }
        else if let after = number(o["reset_after_seconds"]) { reset = fetchedAt.addingTimeInterval(after) }
        let length = number(o["limit_window_seconds"]).flatMap { $0 > 0 ? $0 : nil }
        return UsageWindow(percent: min(max(pct, 0), 100), resetsAt: reset, length: length)
    }

    private static func window(_ value: Any?, length: TimeInterval) -> UsageWindow? {
        guard let o = value as? [String: Any], let pct = number(o["utilization"]) else { return nil }
        return UsageWindow(percent: min(max(pct, 0), 100), resetsAt: (o["resets_at"] as? String).flatMap(parseDate), length: length)
    }

    private static func parseExtra(_ value: Any?) -> ExtraUsage? {
        guard let o = value as? [String: Any] else { return nil }
        // Credits are reported in cents.
        let limit = number(o["monthly_limit"]).map { $0 / 100 }
        let used = number(o["used_credits"]).map { $0 / 100 }
        var pct = number(o["utilization"])
        if pct == nil, let l = limit, l > 0, let u = used { pct = u / l * 100 }
        return ExtraUsage(enabled: (o["is_enabled"] as? Bool) ?? false, limitUSD: limit, usedUSD: used,
                          percent: pct.map { min(max($0, 0), 100) })
    }

    private static func number(_ v: Any?) -> Double? {
        // JSON booleans also arrive as NSNumber; tell them apart by CF type rather than `is Bool`,
        // which is also true for the numbers 0 and 1.
        guard let n = v as? NSNumber, CFGetTypeID(n) != CFBooleanGetTypeID() else { return nil }
        return n.doubleValue
    }

    /// ISO-8601 with or without (possibly 6-digit) fractional seconds.
    static func parseDate(_ s: String) -> Date? {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let d = f.date(from: s) { return d }
        f.formatOptions = [.withInternetDateTime]
        let trimmed = s.replacingOccurrences(of: #"\.\d+"#, with: "", options: .regularExpression)
        return f.date(from: trimmed)
    }
}

enum FetchStatus: Equatable {
    case none, ok, noCredentials, keychainDenied, unauthorized, rateLimited, error
}

struct FetchResult {
    var status: FetchStatus
    var data: UsageSnapshot?
    var plan: String?
    var message: String?
    var retryAfter: TimeInterval?
}

// MARK: - Credentials

/// Claude Code keeps its login in the macOS Keychain ("Claude Code-credentials"); older installs used a file.
enum Credentials {
    enum Result {
        case found(token: String, plan: String?)
        case missing(plan: String?)
        case denied
    }

    static var configDir: URL {
        if let d = ProcessInfo.processInfo.environment["CLAUDE_CONFIG_DIR"], !d.isEmpty {
            return URL(fileURLWithPath: d)
        }
        return FileManager.default.homeDirectoryForCurrentUser.appendingPathComponent(".claude")
    }

    /// Reading the Keychain item shows a macOS permission prompt the first time; "Always Allow" makes it silent.
    static func read() -> Result {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: "Claude Code-credentials",
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
        ]
        var out: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &out)
        switch status {
        case errSecSuccess:
            if let data = out as? Data, let r = parse(data), case .found = r { return r }
        case errSecUserCanceled, errSecAuthFailed, errSecInteractionNotAllowed:
            return .denied
        default:
            break
        }

        let file = configDir.appendingPathComponent(".credentials.json")
        if let data = try? Data(contentsOf: file), let r = parse(data) { return r }
        return .missing(plan: nil)
    }

    private static func parse(_ data: Data) -> Result? {
        guard let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let o = root["claudeAiOauth"] as? [String: Any] else { return nil }
        let plan = o["subscriptionType"] as? String
        if let token = o["accessToken"] as? String, !token.isEmpty { return .found(token: token, plan: plan) }
        return .missing(plan: plan)
    }
}

/// Codex keeps its ChatGPT login in ~/.codex/auth.json (or $CODEX_HOME/auth.json).
enum CodexCredentials {
    static var home: URL {
        if let d = ProcessInfo.processInfo.environment["CODEX_HOME"], !d.isEmpty {
            return URL(fileURLWithPath: d)
        }
        return FileManager.default.homeDirectoryForCurrentUser.appendingPathComponent(".codex")
    }

    static func read() -> (token: String, account: String?)? {
        guard let data = try? Data(contentsOf: home.appendingPathComponent("auth.json")),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let tokens = root["tokens"] as? [String: Any],
              let token = tokens["access_token"] as? String, !token.isEmpty else { return nil }
        return (token, tokens["account_id"] as? String)
    }
}

// MARK: - API client

/// Refuses redirects so the bearer token can only ever be sent to the provider's own host.
private final class NoRedirect: NSObject, URLSessionTaskDelegate {
    func urlSession(_ session: URLSession, task: URLSessionTask, willPerformHTTPRedirection response: HTTPURLResponse,
                    newRequest request: URLRequest, completionHandler: @escaping (URLRequest?) -> Void) {
        completionHandler(nil)
    }
}

/// Fetches plan usage with a CLI's own login: Claude Code's for Claude, Codex's for ChatGPT.
final class UsageClient {
    let provider: Provider
    private let session: URLSession
    private var token: String?
    private var account: String?
    private var plan: String?
    private var rejected: String?

    private var endpoint: URL {
        URL(string: provider == .chatgpt ? "https://chatgpt.com/backend-api/wham/usage" : "https://api.anthropic.com/api/oauth/usage")!
    }
    private var host: String { provider == .chatgpt ? "chatgpt.com" : "api.anthropic.com" }
    private var company: String { provider == .chatgpt ? "OpenAI" : "Anthropic" }

    init(provider: Provider) {
        self.provider = provider
        let c = URLSessionConfiguration.ephemeral
        c.timeoutIntervalForRequest = 20
        c.httpCookieStorage = nil
        c.urlCache = nil
        session = URLSession(configuration: c, delegate: NoRedirect(), delegateQueue: nil)
    }

    /// Forget the cached token so the next fetch re-reads the Keychain (e.g. after the user allows access).
    func reset() { token = nil }

    func fetch() async -> FetchResult {
        if provider == .chatgpt {
            // Codex's login is a plain file, so it's re-read every time and a new login is picked up straight away.
            let creds = CodexCredentials.read()
            token = creds?.token
            account = creds?.account
        } else if token == nil || token == rejected {
            // The token is kept in memory so the Keychain is only read at start-up and after a rejection.
            switch Credentials.read() {
            case .found(let t, let p): token = t; plan = p
            case .missing(let p): token = nil; plan = p ?? plan
            case .denied: token = nil
                return FetchResult(status: .keychainDenied, plan: plan, message: "Keychain access was denied")
            }
        }
        guard let token else {
            return FetchResult(status: .noCredentials, plan: plan, message: "Not signed in")
        }
        if token == rejected {
            return FetchResult(status: .unauthorized, plan: plan, message: "Login expired")
        }

        var req = URLRequest(url: endpoint)
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        if provider == .chatgpt {
            if let account, !account.isEmpty { req.setValue(account, forHTTPHeaderField: "ChatGPT-Account-Id") }
        } else {
            req.setValue("oauth-2025-04-20", forHTTPHeaderField: "anthropic-beta")
        }
        req.setValue("ClaudeUsageBar-macOS/1.0", forHTTPHeaderField: "User-Agent")

        do {
            let (data, response) = try await session.data(for: req)
            guard let http = response as? HTTPURLResponse else {
                return FetchResult(status: .error, plan: plan, message: "Couldn't update")
            }
            switch http.statusCode {
            case 200..<300:
                guard data.count < 256 * 1024,
                      let root = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                    return FetchResult(status: .error, plan: plan, message: "Unexpected response")
                }
                let now = Date()
                if let p = root["plan_type"] as? String, !p.isEmpty { plan = p }
                Cache.save(raw: root, plan: plan, at: now, provider: provider)
                return FetchResult(status: .ok, data: UsageSnapshot.parse(root, provider: provider, fetchedAt: now), plan: plan)
            case 429:
                let wait = http.value(forHTTPHeaderField: "Retry-After").flatMap { TimeInterval($0) }
                return FetchResult(status: .rateLimited, plan: plan, message: "Rate limited by \(company)", retryAfter: wait)
            case 401, 403:
                rejected = token
                return FetchResult(status: .unauthorized, plan: plan, message: "Login expired")
            default:
                return FetchResult(status: .error, plan: plan, message: "Server error (\(http.statusCode))")
            }
        } catch let e as URLError where e.code == .timedOut {
            return FetchResult(status: .error, plan: plan, message: "Request timed out")
        } catch is URLError {
            return FetchResult(status: .error, plan: plan, message: "Can't reach \(host)")
        } catch {
            return FetchResult(status: .error, plan: plan, message: "Couldn't update")
        }
    }
}

// MARK: - Cache (last usage numbers only — never the token)

enum Cache {
    static func url(_ provider: Provider) -> URL {
        let dir = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("ClaudeUsageBar", isDirectory: true)
        return dir.appendingPathComponent(provider == .chatgpt ? "last-chatgpt.json" : "last.json")
    }

    static func save(raw: [String: Any], plan: String?, at: Date, provider: Provider) {
        let file = url(provider)
        var obj: [String: Any] = ["fetchedAt": at.timeIntervalSince1970, "data": raw]
        if let plan { obj["plan"] = plan }
        guard let data = try? JSONSerialization.data(withJSONObject: obj) else { return }
        try? FileManager.default.createDirectory(at: file.deletingLastPathComponent(), withIntermediateDirectories: true)
        try? data.write(to: file, options: .atomic)
    }

    static func load(_ provider: Provider) -> (data: UsageSnapshot, plan: String?)? {
        guard let data = try? Data(contentsOf: url(provider)),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let at = obj["fetchedAt"] as? Double,
              let raw = obj["data"] as? [String: Any] else { return nil }
        return (UsageSnapshot.parse(raw, provider: provider, fetchedAt: Date(timeIntervalSince1970: at)), obj["plan"] as? String)
    }
}

// MARK: - Formatting

enum Fmt {
    static let fiveHours: TimeInterval = 5 * 3600
    static let sevenDays: TimeInterval = 7 * 86400

    /// Compact countdown for the menu bar: 42m, 3h 5m, 4d 3h.
    static func short(_ t: TimeInterval) -> String {
        if t <= 0 { return "now" }
        let mins = Int((t / 60).rounded(.up))
        if t < 3600 { return "\(max(1, mins))m" }
        let h = Int(t / 3600), m = Int(t.truncatingRemainder(dividingBy: 3600) / 60)
        if t < 86400 { return m == 0 ? "\(h)h" : "\(h)h \(m)m" }
        let d = Int(t / 86400), dh = Int(t.truncatingRemainder(dividingBy: 86400) / 3600)
        return dh == 0 ? "\(d)d" : "\(d)d \(dh)h"
    }

    static func reset(_ date: Date?, now: Date) -> String {
        guard let date else { return "" }
        let left = date.timeIntervalSince(now)
        if left <= 0 { return "Resetting…" }
        if left < 3600 { return "Resets in \(max(1, Int((left / 60).rounded(.up))))m" }
        if left < 86400 {
            return String(format: "Resets in %dh %02dm", Int(left / 3600), Int(left.truncatingRemainder(dividingBy: 3600) / 60))
        }
        let f = DateFormatter()
        f.setLocalizedDateFormatFromTemplate("EEE j:mm")
        return "Resets " + f.string(from: date)
    }

    static func ago(_ t: TimeInterval) -> String {
        if t < 45 { return "just now" }
        if t < 3600 { return "\(max(1, Int((t / 60).rounded()))) min ago" }
        if t < 86400 { return "\(Int(t / 3600)) h ago" }
        return "\(Int(t / 86400)) d ago"
    }

    static func inTime(_ t: TimeInterval) -> String {
        t < 60 ? "\(max(1, Int(t.rounded(.up))))s" : "\(Int((t / 60).rounded(.up))) min"
    }

    static func money(_ v: Double) -> String {
        v >= 1000 ? String(format: "$%.0f", v) : String(format: "$%.2f", v)
    }

    static func plan(_ p: String?) -> String {
        switch p?.lowercased() {
        case nil, "": return ""
        case "pro": return "Pro"
        case "plus": return "Plus"
        case "max": return "Max"
        case "team": return "Team"
        case "enterprise": return "Enterprise"
        case let other?: return other.capitalized
        }
    }

    /// Short menu bar label for a window: "5h", "7d". Falls back when the length is unknown.
    static func label(_ w: UsageWindow?, fallback: String) -> String {
        guard let l = w?.length else { return fallback }
        if l >= 86400 { return "\(Int((l / 86400).rounded()))d" }
        if l >= 3600 { return "\(Int((l / 3600).rounded()))h" }
        return fallback
    }

    /// "5-hour rolling window", "Weekly window": describes a window by its length.
    static func describe(_ length: TimeInterval) -> String {
        if abs(length - sevenDays) < 43200 { return "Weekly window" }
        if length >= 86400 { return "\(Int((length / 86400).rounded()))-day window" }
        return "\(Int((length / 3600).rounded()))-hour rolling window"
    }

    /// How far through the window we are (0...1), for the pace marker.
    static func elapsed(_ reset: Date?, length: TimeInterval, now: Date) -> Double? {
        guard let reset else { return nil }
        return min(max(1 - reset.timeIntervalSince(now) / length, 0), 1)
    }
}
